crypto/pkcs12/p12_kiss.c

```diff
@@ -49,27 +49,28 @@ int PKCS12_parse(PKCS12 *p12, const char *pass, EVP_PKEY **pkey, X509 **cert,
     }
 
     /* Check the mac */
-
-    /*
-     * If password is zero length or NULL then try verifying both cases to
-     * determine which password is correct. The reason for this is that under
-     * PKCS#12 password based encryption no password and a zero length
-     * password are two different things...
-     */
-
-    if (pass == NULL || *pass == '\0') {
-        if (!PKCS12_mac_present(p12)
-            || PKCS12_verify_mac(p12, NULL, 0))
-            pass = NULL;
-        else if (PKCS12_verify_mac(p12, "", 0))
-            pass = "";
-        else {
+    if (PKCS12_mac_present(p12)) {
+        /*
+         * If password is zero length or NULL then try verifying both cases to
+         * determine which password is correct. The reason for this is that under
+         * PKCS#12 password based encryption no password and a zero length
+         * password are two different things...
+         */
+        if (pass == NULL || *pass == '\0') {
+            if (PKCS12_verify_mac(p12, NULL, 0))
+                pass = NULL;
+            else if (PKCS12_verify_mac(p12, "", 0))
+                pass = "";
+            else {
+                ERR_raise(ERR_LIB_PKCS12, PKCS12_R_MAC_VERIFY_FAILURE);
+                goto err;
+            }
+        } else if (!PKCS12_verify_mac(p12, pass, -1)) {
             ERR_raise(ERR_LIB_PKCS12, PKCS12_R_MAC_VERIFY_FAILURE);
             goto err;
         }
-    } else if (!PKCS12_verify_mac(p12, pass, -1)) {
```

Comment:
Why not keep the original code here and add `PKCS12_mac_present(p12) &&` to this if?

It's a less invasive change that is more likely to be accepted onto stable branches.

Reply:
That would do the trick, I guess. That said, my thinking (certainly influenced by the Solaris work I do) is that backports should be as close to the original change as possible, to avoid divergence (which might impact future backports on top of the current one) and behavioral changes. There is no change of behavior with the alternative approach as far as I can tell so I am happy to do the change if that makes it acceptable as a backport.

Reply:
I could leave out the tests as well, which is a bit disputable, however it would make the change really minimal.

Reply:
I do not think the size of the change is what blocks the merge to stable branches. IMO it is a new feature.

Reply:
Well, it could be viewed as a regression. It certainly is from where I sit, breaking functionality in FIPS mode when transitioning from 1.0 to 3.0. On the other hand I have to say that this argument is not very strong since it has other implications/context.

Reply:
> It certainly is from where I sit, breaking functionality in FIPS mode when transitioning from 1.0 to 3.0.

The new FIPS module is not backwards compatible with the old module and there was no attempt to make it so. It was a complete rewrite from scratch. So this isn't really a regression.